metabase/metabase
# fix(sdk): font size, color and padding in viz

Closes https://github.com/metabase/metabase/issues/44222 and https://github.com/metabase/metabase/issues/44226

### Description

Fixes the font sizes, colors and padding in visualizations for the embedding environment:

- Remove the hard-coded `isEmbeddingSdk` check and make chart padding customizable
   - Set card padding default to "0.5rem 1rem", the same as in Metabase app. The SDK default is applied before the `isQueryBuilder` check so the SDK padding is applied to Interactive Questions as well.
- Replace `px` and `rem` with `em` to enable font size scaling, i.e. configuring `theme.fontSize` to take effect. This is measured to not cause any noticeable changes in the main Metabase app.
   - Affected elements: pivot table cell, funnel typography, gauge label, chart legend item title.
   - How to test: set the font size to "0px" in the demo app. It should disappear if it's affected by font size scaling.
   - I have double-checked both dashboards and questions in the main Metabase app to ensure that the use of `em` here doesn't cause any cascading issues. When they do, we use the `units(...)` mechanism instead to retain `px` in the Metabase app.
- Change funnel text color to `text-dark` so it is themeable.

### Known exceptions to font size scaling

Changing `theme.fontSize` won't affect any of these elements:

- Data point values in cartesian charts (aka the value shown above bar / lines)
  - Trivial, but will cause this PR to be too big. Let's do it later.
- Previous value comparison text in smart scalar
  - Causes issue with text measurements, in which the breakpoint where the previous value comparison is visible changes quite a bit. Might be good to add an option later to hide the previous value comparison. I'll leave this out of the PR for now 😞 

### Caveat

- The trade-off to using `em` as opposed to `rem` is that it is affected by parent's font sizes, which cascades down the child. This is why I try to retain `px` for the main Metabase app for cases where it's not reliable to use `em`.
  - I've tested the changes in the Question Editor and Dashboard Editor, trying to identify potential cascade issues. Where else should I be testing in?

### How to verify

- Ensure that in the main Metabase app, there are no noticeable font size changes. This is already partially ensured by the use of Loki visual tests.
- Go to the `test-remaining-viz` branch in the demo app.
- Navigate to `/dev` to go to the Kitchen Sink.
- Change `theme.fontSize` in `AppProvider` to your desired values, e.g. `0px`, `14px`, `16px`, `18px`.
- The typography of all chart elements should change accordingly.

### Demo

Metabase app remains as-is, nothing changes

![CleanShot 2567-06-17 at 21 30 00@2x](https://github.com/metabase/metabase/assets/4714175/2c504405-a581-440d-907b-867e1fe3fad3)

Setting `theme.fontSize` to `0px` should remove all text in the viz test

![CleanShot 2567-06-17 at 21 31 06@2x](https://github.com/metabase/metabase/assets/4714175/b0a7c780-5d95-424b-9ece-4febe1783b72)

Setting `theme.fontSize` to `16px` should make everything bigger

![CleanShot 2567-06-17 at 21 46 31@2x](https://github.com/metabase/metabase/assets/4714175/715b488c-416f-45a5-a4b9-770a3d3ce643)

Setting `theme.components.cartesian.padding` to `2px 4px` to reduce the padding around charts

![CleanShot 2567-06-17 at 22 17 08@2x](https://github.com/metabase/metabase/assets/4714175/8360f5dd-4587-4c92-afcf-56ee6d7b7afc)

Setting `theme.components.cartesian.padding` to `0.5rem 1rem`, which is the new default. You can also remove the `padding` theme option from `AppProvider` to get this.

![CleanShot 2567-06-17 at 22 16 57@2x](https://github.com/metabase/metabase/assets/4714175/5d2e9bd8-3f13-4ca9-a4c5-befc7fbc49d9)

### Checklist

- [x] Tests have been added/updated to cover changes in this PR - visual test
